skyportal/handlers/api/listing.py

```diff
@@ -0,0 +1,359 @@
+        parameters:
+        - in: path
+          name: user_id
+          required: true
+          type: string
+        - in: query
+          name: list_name
+          required: false
+          type: string
+
+        description: |
+            find all objects saved to this list.
+            If not given will return all objects
+            saved by the user to all lists.
+
+        responses:
+          200:
+            content:
+              schema: ArrayOFListings
+        """
+
+        if User.query.get(user_id) is None:  # verify that user exists
+            return self.error(f'User "{user_id}" does not exist!')
+
+        user_id = int(user_id)
+
+        # verify that poster has read access to user_id's lists
+        if (
+            self.associated_user_object.id != user_id
+            and "System admin" not in self.associated_user_object.permissions
+        ):
+            return self.error('Insufficient permissions to access this listing. ')
+
+        list_name = self.get_query_argument("list_name", None)
+
+        objects = DBSession().query(Listing).filter(Listing.user_id == user_id)
+
+        if list_name is not None:
+            objects = objects.filter(Listing.list_name == list_name)
+            if objects.first() is None:
+                return self.error(f'No entries under the "{list_name}" list. ')
+
+        return self.success(data=objects.all())
+
+    def add_listing(self, error_if_exists):
+        """
+        Add the listing given in the request body.
+        Two different behaviors when error_if_exists is true/false:
+        - true: will throw an error if listing exists (POST)
+        - false: will ignore if listing already exists (PUT)
+        """
+
+        data = self.get_json()
+
+        schema = Listing.__schema__()
+        try:
+            schema.load(data)
+        except ValidationError as e:
+            return self.error(f'Invalid/missing parameters: {e.normalized_messages()}')
+
+        user_id = data.get('user_id')
+
+        if User.query.get(user_id) is None:  # verify that user exists
+            return self.error(f'User "{user_id}" does not exist!')
+
+        user_id = int(user_id)
+
+        # verify that poster has write access to user_id's lists
+        if (
+            not self.associated_user_object.id == user_id
+            and "System admin" not in self.associated_user_object.permissions
+        ):
+            return self.error('Insufficient permissions to access this listing. ')
+
+        obj_id = data.get('obj_id')
+        if Obj.query.get(obj_id) is None:  # verify that object exists!
+            return self.error(f'Object "{obj_id}" does not exist!')
+
+        list_name = data.get('list_name')
+        if not re.search(r'^\w+', list_name):
+            return self.error(
+                "Input `list_name` must begin with alphanumeric/underscore"
+            )
+
+        objects = (
```

Comment:
```suggestion
        query = (
```
and throughout below